packages/services/test/src/utils.ts

```diff
@@ -454,3 +679,119 @@ class TerminalTester extends SocketTester {
+ * giving a test function. And you can test the first signal matching the
+ * find criteria by giving both.
+ */
+export
+async function testEmission<T, U, V>(signal: ISignal<T, U>, options: {
+  find?: (a: T, b: U) => boolean
+  test?: (a: T, b: U) => void,
+  value?: V
+}): Promise<V> {
+  const done = new PromiseDelegate<V>();
+  const object = {};
+  signal.connect((sender: T, args: U) => {
+    if (!options.find || options.find(sender, args)) {
+      try {
+        Signal.disconnectReceiver(object);
+        if (options.test) {
+          options.test(sender, args);
+        }
+      } catch (e) {
+        done.reject(e);
+        // throw e;
+      }
+      done.resolve(options.value || undefined);
+    }
+  }, object);
+  return done.promise;
+}
+
+/**
+ * Test that promises resolve in order.
+ *
+ * @param promises - a list of promises in the order they should resolve
+ * @param resolutions - a list of resolved values expected from the promises
+ *
+ * #### Notes
+ * The promises must resolve to different values (i.e., resolutions must be a
+ * list of unique values) for this function to work.
+ */
+export
+function testResolveOrder<T1, T2, T3, T4, T5, T6, T7>(promises: [PromiseLike<T1>, PromiseLike<T2>, PromiseLike<T3>, PromiseLike<T4>, PromiseLike<T5>, PromiseLike<T6>, PromiseLike<T6>], resolutions: [T1, T2, T3, T4, T5, T6, T7]): Promise<void>;
+export
+function testResolveOrder<T1, T2, T3, T4, T5, T6>(promises: [PromiseLike<T1>, PromiseLike<T2>, PromiseLike<T3>, PromiseLike<T4>, PromiseLike<T5>, PromiseLike<T6>], resolutions: [T1, T2, T3, T4, T5, T6]): Promise<void>;
+export
+function testResolveOrder<T1, T2, T3, T4, T5>(promises: [PromiseLike<T1>, PromiseLike<T2>, PromiseLike<T3>, PromiseLike<T4>, PromiseLike<T5>], resolutions: [T1, T2, T3, T4, T5]): Promise<void>;
+export
+function testResolveOrder<T1, T2, T3, T4>(promises: [PromiseLike<T1>, PromiseLike<T2>, PromiseLike<T3>, PromiseLike<T4>], resolutions: [T1, T2, T3, T4]): Promise<void>;
+export
+function testResolveOrder<T1, T2, T3>(promises: [PromiseLike<T1>, PromiseLike<T2>, PromiseLike<T3>], resolutions: [T1, T2, T3]): Promise<void>;
+export
+async function testResolveOrder<T1, T2>(promises: [PromiseLike<T1>, PromiseLike<T2>], resolutions: [T1, T2]): Promise<void>;
+export
+async function testResolveOrder(promises: PromiseLike<any>[], resolutions: any[]): Promise<void> {
+  // We construct all of the races synchronously so that the winner can truly be determined
+  let subsequences = promises.map((value, index) => Promise.race(promises.slice(index)));
+  for (let i = 0; i < subsequences.length; i++) {
+    expect(await subsequences[i]).to.be(resolutions[i]);
+  }
+}
+
+/**
+ * Test to see if a promise is fulfilled.
+ *
+ * @returns true if the promise is fulfilled (either resolved or rejected), and
+ * false if the promise is still pending.
+ */
+export
+async function isFulfilled<T>(p: PromiseLike<T>): Promise<boolean> {
```

Comment:
Very clever. I had thought there was no way to do this with JS. At least it's not synchronous, or else my whole world view would be challenged.